### PR TITLE
Directions v5 

### LINKED ISSUE #120: Directions API update
@perrygeo let's coordinate a new release to go along with the next Directions API update.

### PR DESCRIPTION
**Pull request by perrygeo**
This PR updates the `directions` module to version v5 of the Mapbox HTTP API.

From the perspective of the Python SDK user, I *think* there are only two backwards-incompatible changes: 

* the `instructions` argument is no longer supported in v5. Currently, the `directions` method 
 will silently ignored it if provided, TODO: something better.

* The returned geometry in the `json` payload is polyline by default in v5. 

The underlying implementation has changed significantly enough to cause a large diff. I'll try to summarize the major changes here:

* The `geojson()` method has to decode the default polyline geometries into geojson geometries.

* The `geometry` parameter is replaced by `geometries` in v5. The old parameter will work but raises a deprecation warning.

* new format for profiles; `mapbox.cycling` in `mapbox/cycling` v4. The old style will still work but is deprecated.

* Several new arguments added in v5. (see https://www.mapbox.com/api-documentation/#retrieve-directions). The validation of these params makes up the bulk of the changeset.

* docstrings, still wavering on how much detail to copy from HTTP api docs into this docstring.

Some things I'm not comfortable with yet: 

* v5 gives users additional control over how input coordinates are snapped to roads (see the `radiuses` and `bearings` arguments). I'm struggling with the fact that both of those arguments need to be lists of the same length as `features`. They are effectively properties of the input features but need to be provided separately. It feels strange. Maybe parsing these optional properties from the input GeoJSON Features would be better? 

In terms of supporting v4, I decided against it because we can implement most of the SDK behavior on top of the v5 API. The only thing missing is the textual instructions; *If* we need to use the v4 implementation to support this use case then I'd suggest making it an entirely separate service class. Mixing the code and tests would cause way too much branching headache.

@sgillies This is still a work in progress but ready for preliminary review.

Resolves #120

### CODE REVIEW

File: mapbox/services/directions.py [hunk 1]
```diff
@@ -1,103 +1,223 @@
...
+            where the bearing tuple consists of an angle and range
+                [(215, 45), (315, 90)]
+            Must be same length as features
+        annotations: string
+        language: string
+
+        Returns
+        -------
+        requests.Response
+            the json() method will return a Directions response dict
```
Comment (sgillies) [inline mapbox/services/directions.py]: @perrygeo `bearings` and `radius` are optional, but have to be paired up when present, yes? How about an argument that is a sequence of (bearing, radius) pairs? That would let us eliminate some validation from this method and would make the requirement that the lists be zippable very explicit.

Separate `feature` and `bearings/radii` args (maybe `(angle, range, radius)` triplets?) make sense to me. The former are require, the latter are optional. As long as we're clear to users that the method will zip them, we should be okay.

Comment (perrygeo) [inline mapbox/services/directions.py]: Combining all the waypoint snapping into a single argument makes sense from a Python usability perspective. My only (minor) concern is that we're deviating from the API docs to achieve it. I'm ok with that in this case.

@sgillies any preference for the name of this combined option? `radius_bearings`? `waypoint_snapping`?

Comment (perrygeo) [inline mapbox/services/directions.py]: I added a combined `waypoint_snapping` param in the latest commit (f7e91e9)

The parameter is a bit loose with types and there are a number of potential combos of inputs. From the docstring:

```
waypoint_snapping: list
    List must be same length as features.
    Elements of the list must be one of:
    - A number (interpretted as a snapping radius)
    - The string 'unlimited' (unlimited snapping radius)
    - A 3-element tuple consisting of (radius, angle, range)
    - None (no snapping parameters specified for that waypoint)
```

Makes internal validation code a little complex but IMO improves usability.

**Comment by sgillies** [comment]
Our upload API doctest failed for what looks like the usual reasons. I'd double check if this was a release, but will merge this despite the red light.

**Pull request merged**